File: week_6/pset_6/sentimental-credit/credit.py
from cs50 import get_int


def card_type(card):

    card = str(card)

    # Check what type of credit card is or if its invalid
    if (card[0] == '4' and (len(card) == 13 or len(card) == 16)):  # Check if its a VISA
        print("VISA\n")
    elif (card[:2] == '34' or card[:2] == '37' and len(card) == 15):  # Check if its a AMERICAN EXPRESS
        print("AMEX\n")
    elif (int(card[:2]) >= 51 and int(card[:2]) <= 55 and len(card) == 16):  # Check if its a MASTERCARD
        print("MASTERCARD\n")
    else:
        print("INVALID\n")


def check_sum(card):

    card = str(card)

    # Checks if the credit card meets Luhn's algorithm
    sum_p = 0
    sum_o = 0

    card_copy = card[::-1]

    # Sum of every pair digit multipliyed by 2
    for i in card_copy[1:len(card_copy):2]:
        digit = int(i) * 2
        if digit >= 10:
            digit1 = str(digit)
            digit = int(digit1[0]) + int(digit1[1])
        sum_p += digit

    for i in card_copy[0:len(card_copy):2]:
        sum_o += int(i)

    sum = sum_p + sum_o

    if (str(sum)[-1] == '0'):
        return True
    else:
        return False


def main():

    credit_card = get_int("Credit Card Number: ")

    # Conditional to solve the problem
    if (check_sum(credit_card) == True):
        card_type(credit_card)
    else:
        print("INVALID")


if __name__ == '__main__':
    main()

File: week_1/pset_1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

void card_type(long card);
bool check_sum(long card);

int main(void)
{
    long credit_card = get_long("Credit Card Number: ");

    //Conditional to solve the problem
    if (check_sum(credit_card) == true)
    {
        card_type(credit_card);
    }
    else
    {
        printf("INVALID\n");
    }
}

void card_type(long card)
{
    //Check what type of credit card is or if its invalid
    if (card / 1000000000000 == 4 || card / 1000000000000000 == 4) //Check if it's a VISA
    {
        printf("VISA\n");
    }
    else if (card / 10000000000000 == 34 || card / 10000000000000 == 37) //Check if it's a AMERICAN EXPRESS
    {
        printf("AMEX\n");
    }
    else if (card / 100000000000000 >= 51 && card / 100000000000000 <= 55) //Check if it's a MASTERCARD
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

bool check_sum(long card)
{
    //Checks if the credit card meets Luhn's algorithm
    int pairs = 0;
    int odds = 0;

    //Sum of every pair digit multipliyed by 2
    int d2 = (((card / 10) % 10) * 2 / 10) + (((card / 10 % 10) * 2) % 10);
    int d4 = (((card / 1000) % 10) * 2 / 10) + (((card / 1000 % 10) * 2) % 10);
    int d6 = (((card / 100000) % 10) * 2 / 10) + (((card / 100000 % 10) * 2) % 10);
    int d8 = (((card / 10000000) % 10) * 2 / 10) + (((card / 10000000 % 10) * 2) % 10);
    int d10 = (((card / 1000000000) % 10) * 2 / 10) + (((card / 1000000000 % 10) * 2) % 10);
    int d12 = (((card / 100000000000 % 10) * 2) / 10) + (((card / 100000000000 % 10) * 2) % 10);
    int d14 = (((card / 10000000000000) % 10) * 2 / 10) + (((card / 10000000000000 % 10) * 2) % 10);
    int d16 = (((card / 1000000000000000) % 10) * 2 / 10) + (((card / 1000000000000000 % 10) * 2) % 10);

    //Sum of every odd digit
    int d1 = (card % 10);
    int d3 = ((card / 100) % 10);
    int d5 = ((card / 10000) % 10);
    int d7 = ((card / 1000000) % 10);
    int d9 = ((card / 100000000) % 10);
    int d11 = ((card / 10000000000) % 10);
    int d13 = ((card / 1000000000000) % 10);
    int d15 = ((card / 100000000000000) % 10);

    int sum_p = d2 + d4 + d6 + d8 + d10 + d12 + d14 + d16;
    int sum_o = d1 + d3 + d5 + d7 + d9 + d11 + d13 + d15;
    int sum = sum_p + sum_o;

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }

}